File: Firmware/RawMessageUnpacker.cpp
#include "RawMessageUnpacker.h"


#define StartByte 0xFF


bool RawMessageUnpacker::Unpack(uint8_t data, RawMessage &message)
{
  return Process(data, message);
}


bool RawMessageUnpacker::Process(uint8_t data, RawMessage &message)
{
    switch (dataSize)
    {
        case 0:
            ProcessStart(data);
            return false;
        case 1:
            ProcessId(data);
            return false;
        case 2:
            ProcessLength(data);
            return false;
        case 3:
            ProcessChecksum(data);
            return false;
        default:
            return ProcessPayload(data, message);
            break;
    }
}


void RawMessageUnpacker::ProcessStart(uint8_t data)
{
    AddByte(data);

    if (data != StartByte)
    {
        Abort();
    }
}


void RawMessageUnpacker::ProcessId(uint8_t data)
{
    AddByte(data);

    id = data;
}


void RawMessageUnpacker::ProcessLength(uint8_t data)
{
    AddByte(data);

    length = data;
    //payloadData = new byte[length];
}


void RawMessageUnpacker::ProcessChecksum(uint8_t data)
{
    AddByte(data);

    checksum = data;
}


bool RawMessageUnpacker::ProcessPayload(uint8_t data, RawMessage &message)
{
    // TODO: If length==0 then this code will still consume one byte as payload
    AddByte(data);

    auto payloadSize = dataSize - 4;
    payloadData[payloadSize - 1] = data;

    if (payloadSize >= length)
    {
        // Verify checksum
        auto expectedChecksum = CalcChecksum(payloadData, payloadSize);
        if (checksum != expectedChecksum)
        {
            Abort();
            return false;
        }

        // We are done!
        message.Setup(id, payloadData, payloadSize);

        Reset();
        return true;
    }

    return false;
}


void RawMessageUnpacker::AddByte(uint8_t data)
{
    dataSize++;
    //data[dataSize++] = data;
}


void RawMessageUnpacker::Abort()
{
    // TODO: Log discarded data?

    Reset();
}

void RawMessageUnpacker::Reset()
{
    dataSize = 0;
    //payloadData = null;
}

File: GUI/WpfMidiOrganPedals/Devices/RawMessageUnpacker.cs
using System;

namespace WpfMidiOrganPedals.Devices
{
    internal class RawMessageUnpacker : RawMessageHandler
    {
        private readonly byte[] data;
        private int dataSize;
        private byte id;
        private byte length;
        private byte checksum;
        private byte[] payloadData;

        public RawMessageUnpacker()
        {
            data = new byte[4 + 255]; // 4 fixed bytes and a variable field specified with an 8-bit length

            Reset();
        }

        public Action<RawMessage> MessageFound { get; internal set; }

        internal void Process(byte[] data)
        {
            var index = 0;
            foreach (var @byte in data)
            {
                Process(@byte);
                index++;
            }
        }

        private void Process(byte @byte)
        {
            switch (dataSize)
            {
                case 0:
                    ProcessStart(@byte);
                    break;
                case 1:
                    ProcessId(@byte);
                    break;
                case 2:
                    ProcessLength(@byte);
                    break;
                case 3:
                    ProcessChecksum(@byte);
                    break;
                default:
                    ProcessPayload(@byte);
                    break;
            }
        }

        private void ProcessStart(byte @byte)
        {
            AddByte(@byte);

            if (@byte != StartByte)
            {
                Abort();
            }
        }

        private void ProcessId(byte @byte)
        {
            AddByte(@byte);

            id = @byte;
        }

        private void ProcessLength(byte @byte)
        {
            AddByte(@byte);

            length = @byte;
            payloadData = new byte[length];
        }

        private void ProcessChecksum(byte @byte)
        {
            AddByte(@byte);

            checksum = @byte;
        }

        private void ProcessPayload(byte @byte)
        {
            // TODO: If length==0 then this code will still consume one byte as payload
            AddByte(@byte);

            var payloadSize = dataSize - 4;
            payloadData[payloadSize-1] = @byte;

            if (payloadSize >= length)
            {
                // Verify checksum
                var expectedChecksum = CalcChecksum(payloadData);
                if (checksum != expectedChecksum)
                {
                    Abort();
                    return;
                }

                // We are done!
                var message = new RawMessage(id, payloadData);
                NotifyMessageFound(message);

                Reset();
            }
        }

        private void AddByte(byte @byte)
        {
            data[dataSize++] = @byte;
        }

        private void Abort()
        {
            // TODO: Log discarded data?

            Reset();
        }

        private void Reset()
        {
            dataSize = 0;
            payloadData = null;
        }

        private void NotifyMessageFound(RawMessage message)
        {
            MessageFound.Invoke(message);
        }
    }
}
